### PR TITLE
Adds PyTorch DLRM Workload

### PR DESCRIPTION
**Pull request by rakshithvasudev**
This is a Preliminary PR for Pytorch DLRM workload. 

**Still trying to sort out the deadlock issue. But here is what I have so far:**

Reference model implementation ported from JAX
Dataloader working fast enough 
Basic setup of optimizer inplace - might need tweaking and equivalence checking to JAX version 
Has DDP support 
I have some scripts testing the forward prop and reverse prop time, workload takes about 30 ish mins to train with the full dataset. 
Still doesn't converge to target AUC. 

Once I sort the deadlock issue, I can get to the convergence issue.
Data loader loads the whole data in 600 ish seconds. Just iteration takes 200 - 600s (NVIDIA recommends to place this dataset on a fast storage). 

**TODO: **

- [ ] Add readme that shows preprocessing steps (this preprocessing takes several days - I have it still running from over 4 days now) maybe accelerate this preprocessing step. 
- [ ] Add test cases for equivalence 
- [ ] Improve data loading if needed by splitting the large dataset into shards
- [ ] Remove AUC from metrics if not needed and ensure the target metric is finalized to convergence
- [ ] There are certainly opportunities to improve the training time, will focus once the model is ready and at functional stage

### CODE REVIEW

**Comment by rakshithvasudev** [comment]
I noticed I have some print statements I used for debugging here. Please ignore them. I'll remove them in the upcoming PRs.

Example: 

https://github.com/rakshithvasudev/algorithmic-efficiency/blob/989ae3eb9bcd53fdf36dbee31977ba4dac279098/algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py#L199

**Review by runame — CHANGES_REQUESTED** [review]
Thanks for the PR! I have written a few comments below.

Some more things which should be straight-forward to address:
- Add model/workload to `tests/test_num_params.py` and `tests/test_param_shapes.py`.
- Add `__init__.py` files in all subdirectories of `criteo1tb`.
- Run `isort .` inside of `algorithmic_efficiency` repo.
- Run `yapf -r -i algorithmic_efficiency/`.

I'll do another review after the next iteration of changes.

**Comment by rakshithvasudev** [comment]
Thanks for reviewing the PR @runame. Appreciate the detailed review. Will do these changes. I'd put second priority to these cosmetic changes until I resolve the deadlock issue and convergence issue.  

Great pointers on the convergence issue, will test them and see :) 

Please ignore the inconsistencies with the numbers like batch size and eval time, etc. from the JAX workload not being replicated here. Those are some knobs I was tweaking to try to debug the deadlock.


Thanks again for the PR review! Will push another one soon.

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/input_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+def cycle(iterable, keys=('inputs', 'targets'), custom_sampler=False):
+  iterator = iter(iterable)
+  epoch = 0
+  while True:
+    try:
+      batch = next(iterator)
+      # inputs = numerical_features + categorical_features
+      assert len(keys) + 1 == len(batch)
+      yield {'inputs': (batch[0], batch[1]), 'targets': batch[2]}
```
[the remark this answers is not in the document]
Comment (rakshithvasudev) [reply algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/input_pipeline.py:24]: We need to revisit this once we finalize the dataloading. For now, I've left it as is.

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    data_loader_args = dict(
+        batch_size=None,
+        num_workers=0,
+        pin_memory=False,
+        generator=torch.Generator().manual_seed(int(data_rng[0])),
+        collate_fn=functools.partial(data_collate_fn,
+                                     device=return_device,
+                                     orig_stream=torch.cuda.current_stream()))
+
+    if is_train:
```
[the remark this answers is not in the document]
Comment (rakshithvasudev) [reply algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py:146]: cosmetic, will do.

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+                                              label_batch.reshape(-1, 1),
+                                              reduction='none')
+
+  def _eval_metric(self, logits: spec.Tensor,
+                   targets: spec.Tensor) -> Dict[str, int]:
+
+    # need to remove auc_roc
+    loss = self.loss_fn(logits, targets).sum()
+    auc_roc = metrics.roc_auc_score(torch.sigmoid(logits.float()),
+                                    targets).sum()
```
[the remark this answers is not in the document]
Comment (rakshithvasudev) [reply algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py]: Yes! I tried that. Not sure why github still shows the old code. Here is the updated code: 

https://github.com/rakshithvasudev/algorithmic-efficiency/blob/4c042326fbf2f03681339c84b4e9ad3417c53782/algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py#L334

[the remark this answers is not in the document]
Comment (runame) [reply algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py]: Ah nice, thanks! Ok then it has to be something else, I'll try to take a look at this again tomorrow.

[the remark this answers is not in the document]
Comment (rakshithvasudev) [reply algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py]: Thanks @runame appreciate it. Does a run log help? I can upload a run log. 

It has the loss values and the global steps.

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+      torch.distributed.all_gather(reciever, logits)
+      logits_list.append(output_receive_buffer.float())
+      targets_list.append(batch['targets'])
+    compute_auc = metrics.roc_auc_score(
+        torch.sigmoid(torch.cat(logits_list)), torch.cat(targets_list))
+    if USE_PYTORCH_DDP:
+      for metric in total_metrics.values():
+        dist.all_reduce(metric)
+    computed_val = {
+        k: float(v.item() / num_examples) for k, v in total_metrics.items()
```
Comment (runame) [inline algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py:345]: This is likely not exact, since the last eval batch is dropped, so not all `num_examples` samples are used. But doesn't explain the lack of training performance.

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+  def loss_type(self):
+    return spec.LossType.SIGMOID_CROSS_ENTROPY
+
+  @property
+  def num_train_examples(self):
+    return 4_195_197_692
+
+  @property
+  def num_eval_train_examples(self):
+    return 131072 * 16
```
Comment (runame) [inline algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py:69]: This is `100000` in the Jax workload.

File: algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+
+  @property
+  def num_validation_examples(self):
+    return 131072 * 8  # TODO(znado): finalize the validation split size.
+
+  @property
+  def num_test_examples(self):
+    #return None
+    return 131072 * 16
```
Comment (runame) [inline algorithmic_efficiency/workloads/criteo1tb/criteo1tb_pytorch/workload.py:78]: There is no test set in the Jax workload, i.e. this returns `None`.

**Review by runame — COMMENTED** [review]
Sorry for the late reply. I have left a few comments. Most importantly, DDP/DP does not work with the current code, see my comment in the `model_fn()`. Also, you can try to run it without DDP, to make sure this is not causing any issues.
I haven't found the issue yet just by looking at the code, but you could try to
- match a forward pass with fixed weights and inputs with the Jax model,
- use the same `input_pipeline` as the Jax workload, see the WMT and OGBG workloads for examples on how to do that. What was the reason not to use it in the first place (I can't remember)?

**Comment by znado** [comment]
Closing as this has been subsumed by https://github.com/mlcommons/algorithmic-efficiency/pull/146.
